### PR TITLE
Conway spec prose: GovernanceActions, Voting, Enactment

### PR DESCRIPTION
**Pull request by williamdemeo**
# Description

This addresses three tasks in issue #145.  (GovernanceActions, Voting, and Enactment)

# Checklist

- [X] Commit sequence broadly makes sense and commits have useful messages
- [X] Code is formatted according to [CONTRIBUTING.md](https://github.com/input-output-hk/formal-ledger-specifications/blob/master/CONTRIBUTING.md)
- [X] Self-reviewed the diff

### CODE REVIEW

**Comment by williamdemeo** [comment]
The items in issue #145 that this PR is meant to address says we "should explain the CC, DReps and give a broad overview of the lifetime of a governance action."  Probably this PR is lacking in at least the "broad overview;" also, the explanations of CC and DReps could probably use clarification and maybe should be expanded upon.

File: src/Ledger/GovernanceActions.lagda [hunk 1]
```diff
@@ -112,7 +183,48 @@ NeedsHash (TriggerHF _)        = GovActionID
 NeedsHash (ChangePParams _ _)  = GovActionID
 NeedsHash (TreasuryWdrl _)     = ⊤
 NeedsHash Info                 = ⊤
+\end{code}
 
+A governance action is an on-chain event that is triggered by a transaction, and there is a deadline after which
+the governance action cannot be enacted.
+\begin{itemize}
+\item
+  \defn{Ratification}. An action is said to be \defn{ratified} when it gathers enough votes in its favor (through the rules and parameters that are detailed
```
Comment (WhatisRT) [inline src/Ledger/GovernanceActions.lagda:192]: These can be more specific, we can point to the sections/figures that contain the actual logic.

File: src/Ledger/GovernanceActions.lagda [hunk 2]
```diff
@@ -112,7 +183,48 @@ NeedsHash (TriggerHF _)        = GovActionID
 NeedsHash (ChangePParams _ _)  = GovActionID
 NeedsHash (TreasuryWdrl _)     = ⊤
 NeedsHash Info                 = ⊤
+\end{code}
 
+A governance action is an on-chain event that is triggered by a transaction, and there is a deadline after which
```
Comment (WhatisRT) [inline src/Ledger/GovernanceActions.lagda:188]: Since this document is only about on-chain things we don't need to mention this. Also, an `event` isn't a properly defined thing, so I'd leave that out as well. Here, a governance action is just some data that can be included in a transaction & which can trigger further rules in the ledger.

File: src/Ledger/GovernanceActions.lagda [hunk 3]
```diff
@@ -112,7 +183,48 @@ NeedsHash (TriggerHF _)        = GovActionID
...
+\item
+  \defn{Expiration}. An action that doesn't collect sufficient \AgdaField{yes} votes before its deadline is said to have \defn{expired}.
+\item
+  \defn{Enactment}. An action that has been ratified is said to be \defn{enacted} once it has been activated on the network.
+\end{itemize}
+
+
+\subsection{Voting and ratification}
+\label{sec:voting-and-ratification}
+A governance action is an on-chain event that is triggered by a transaction. There is a deadline after which
```
Comment (WhatisRT) [inline src/Ledger/GovernanceActions.lagda:203]: This sentence and the list below are repeats.

File: src/Ledger/GovernanceActions.lagda [hunk 4]
```diff
@@ -112,7 +183,48 @@ NeedsHash (TriggerHF _)        = GovActionID
...
+  A governance action is said to be \defn{ratified} when it gathers enough votes in its favor
+  (through the rules and parameters that are detailed elsewhere [TODO: insert reference]).
+\item
+  A governance action is said to have \defn{expired} if it does not collect sufficient \AgdaField{yes} votes before the deadline.
+\item
+  An action that has been ratified is said to be \defn{enacted} once it has been activated on the network.
+\end{itemize}
+
+The data type \AgdaDatatype{Vote} represents the different voting options: \AgdaInductiveConstructor{yes}, \AgdaInductiveConstructor{no}, or \AgdaInductiveConstructor{abstain}.
+Each \AgdaField{vote} is recorded in a \AgdaRecord{GovVote} record along with the following meta-data:
```
Comment (WhatisRT) [inline src/Ledger/GovernanceActions.lagda:216]: Only the `anchor` is metadata, everything else is regular data (since we use it in the rules).

File: src/Ledger/GovernanceActions.lagda [hunk 5]
```diff
@@ -131,18 +243,44 @@ record GovProposal : Set where
         prevAction  : NeedsHash action
         anchor      : Anchor
 \end{code}
-\begin{code}[hide]
+\caption{Governance action proposals and votes}
+\label{defs:governance-votes}
+\end{figure*}
+\emph{Any Ada holder} can submit a governance action to the chain.
```
Comment (WhatisRT) [inline src/Ledger/GovernanceActions.lagda:249]: We usually don't mention 'Ada' and 'Lovelace' in the spec. This is also just a marketing line that we don't need to repeat.

File: src/Ledger/GovernanceActions.lagda [hunk 6]
```diff
@@ -178,25 +319,24 @@ private variable
 
 instance
   _ = +-0-monoid
-
-data
```
Comment (WhatisRT) [inline src/Ledger/GovernanceActions.lagda:182]: I'd like to keep this part. It looks a lot like the Shelley spec & hides the type theory. We don't want to gatekeep the spec behind an understanding of 'propositions as types'.

**Review by WhatisRT — CHANGES_REQUESTED** [review]
We should also have a paragraph about `NeedsHash` and `HashProtected`. This is used in ratification for `verifyPrev`, where we check that the stored hash matches the one attached to the action we want to ratify.

We also need to think about how to make the rules for `ENACT` more readable. I find it very hard to parse in its current form.

File: src/Ledger/GovernanceActions.lagda [hunk 7]
```diff
@@ -32,8 +53,53 @@ open import Tactic.Derive.DecEq
 open import MyDebugOptions
 
 2ℚ = 1ℚ Data.Rational.+ 1ℚ
-
 \end{code}
+
+Figure~\ref{defs:governance} defines several data types used to represent governance actions including:
+\begin{itemize}
+  \item %%% \defn{identifier} (\AgdaSymbol{:}\,
+        \AgdaDatatype{GovActionId}---a pair consisting of a transaction id (\AgdaSymbol{:}\,\AgdaDatatype{TxId}) and a natural number (\AgdaSymbol{:}\,\AgdaDatatype{ℕ});
```
[the remark this answers is not in the document]
Comment (williamdemeo) [reply src/Ledger/GovernanceActions.lagda:61]: I removed some of the type annotations and reworded the explanations, referring to the inhabitant of the type instead of the type.

I also rearranged the file so the code hopefully shows up in the pdf in a more appropriate location. Hopefully this improves the flow and makes it more readable.  

Let me know if you think it still needs improvement and maybe specifically how you would suggest improving it.  Thanks!

File: src/Ledger/GovernanceActions.lagda (hunk 1 above)
Comment (williamdemeo) [reply src/Ledger/GovernanceActions.lagda:192]: I added a reference to the separate section/module on Ratification (\ref{sec:ratification}).

File: src/Ledger/GovernanceActions.lagda (hunk 2 above)
Comment (williamdemeo) [reply src/Ledger/GovernanceActions.lagda:188]: Okay, thanks for the clarification.  I'll eliminate "on-chain event."

File: src/Ledger/GovernanceActions.lagda (hunk 5 above)
Comment (williamdemeo) [reply src/Ledger/GovernanceActions.lagda:249]: Okay, I removed references to Ada and Lovelace, but kept the explanation about the deposit (required for submitting action to the chain) that will be refunded later... or did you want me to delete that too?

For the same reason I should remove the paragraph,

"As with stake pools, any Ada holder may register to be a DRep and so choose to represent themselves and/or others. Also, as with stake pools, Ada holders may instead delegate their voting rights to any other DRep. Voting rights will be based on the total Ada that is delegated, as a whole number of Lovelace."

that appears near the start of this section, correct?

**Comment by williamdemeo** [comment]
> We should also have a paragraph about `NeedsHash` and `HashProtected`. This is used in ratification for `verifyPrev`, where we check that the stored hash matches the one attached to the action we want to ratify.
> 
> We also need to think about how to make the rules for `ENACT` more readable. I find it very hard to parse in its current form.

I moved `HashProtected` up so it's defined right after `NeedsHash` and added your explanation below the two definitions.  Feel free to suggest a fuller explanation, or I could read up on those types and write a few more words about them.

**Comment by williamdemeo** [comment]
The forced push noted above just rebases this PR onto the head of recently updated PR #152.

**Comment by WhatisRT** [comment]
I rebased this on `master` so it's easier to review

File: src/Ledger/GovernanceActions.lagda [hunk 8]
```diff
@@ -61,7 +90,37 @@ data GovAction : Set where
   ChangePParams    : UpdateT → PPHash                 → GovAction
   TreasuryWdrl     : (Credential ⇀ Coin)              → GovAction
   Info             :                                    GovAction
-
+\end{code}
+\caption{Governance actions}
+\label{defs:governance}
+\end{figure*}
+\begin{table}[h]
+\begin{longtable}[]{@{}
```
Comment (WhatisRT) [inline src/Ledger/GovernanceActions.lagda:98]: I think this table would look a lot better if the constructor name would go on the left (still under the `Action` heading) and the `Agda type` heading was removed. I don't think we need short and long descriptions of every action in a table.

File: src/Ledger/GovernanceActions.lagda [hunk 9]
```diff
@@ -61,7 +90,37 @@ data GovAction : Set where
...
+3. Updates to the Constitution & A modification to the off-chain Constitution, recorded as an on-chain hash of the text document & \AgdaInductiveConstructor{NewConstitution}\\
+4. Hard fork\footnotemark &
+  Initiation triggers a non-backwards compatible upgrade of the network; requires a prior software upgrade & \AgdaInductiveConstructor{TriggerHF} \\
+5. Protocol Parameter Changes & Any change to \textit{one or more} updatable protocol parameters, excluding changes to major protocol versions (``hard forks'') & \AgdaInductiveConstructor{ChangePParams}\\
+6. Treasury Withdrawals & Movements from the treasury, sub-categorized into small, medium or large withdrawals (based on the amount of Lovelace to be withdrawn). The thresholds for treasury withdrawals are discussed below. & \AgdaInductiveConstructor{TreasuryWdrl} \\
+7. Info & An action that has no effect on-chain, other than an on-chain record. & \AgdaInductiveConstructor{Info}\\
+\hline
+\end{longtable}
+\caption{Types of governance actions}
```
Comment (WhatisRT) [inline src/Ledger/GovernanceActions.lagda:116]: I notice this this is labeled as `Table 2`. Can we just include this in the `figure` counter instead?

File: src/Ledger/GovernanceActions.lagda [hunk 10]
```diff
@@ -131,23 +221,63 @@ record GovProposal : Set where
...
-\end{figure*}
 
+See Section~\ref{sec:ratification} for more on the ratification process.
+
+\subsection{Protocol parameters and governance actions}
+\label{sec:protocol-parameters-and-governance-actions}
+Recall from Section~\ref{sec:protocol-parameters}, the \AgdaRecord{PParam} type is a record containing
+parameters used in the Cardano ledger, grouped according to the general purpose that each parameter serves.
+(See Figure~\ref{fig:protocol-parameter-declarations}.)
+\begin{itemize}
```
Comment (WhatisRT) [inline src/Ledger/GovernanceActions.lagda:254]: We don't need to repeat that list here. The only thing that's relevant in this section is that there are multiple groups, not really their details.

File: src/Ledger/GovernanceActions.lagda [hunk 11]
```diff
@@ -131,23 +221,63 @@ record GovProposal : Set where
...
-\label{defs:governance}
-\end{figure*}
 
+See Section~\ref{sec:ratification} for more on the ratification process.
+
+\subsection{Protocol parameters and governance actions}
+\label{sec:protocol-parameters-and-governance-actions}
+Recall from Section~\ref{sec:protocol-parameters}, the \AgdaRecord{PParam} type is a record containing
+parameters used in the Cardano ledger, grouped according to the general purpose that each parameter serves.
+(See Figure~\ref{fig:protocol-parameter-declarations}.)
```
Comment (WhatisRT) [inline src/Ledger/GovernanceActions.lagda:253]:
```suggestion
(see Figure~\ref{fig:protocol-parameter-declarations}).
```
And the period above should be removed.

File: src/Ledger/GovernanceActions.lagda [hunk 12]
```diff
@@ -131,23 +221,63 @@ record GovProposal : Set where
...
+for multiple parameters from different groups, the maximum threshold of all the groups involved will
+apply to any given such governance action.
+
+\subsection{Enactment}
+\label{sec:enactment}
+\defn{Enactment} of a governance action is carried out as an \defn{enact transition}
+which requires an \defn{enact environment} (\AgdaSymbol{:}\,\AgdaRecord{EnactEnv}), an \defn{enact state}
+(\AgdaSymbol{:}\,\AgdaRecord{EnactState}) representing the existing state (prior to enactment), the voted on
+governance action (that achieved enough votes to enact), and the state that results from enacting the given
+governance action (\AgdaSymbol{:}\,\AgdaRecord{EnactState}). (See Figure~\ref{fig:enactment-types}.)
```
Comment (WhatisRT) [inline src/Ledger/GovernanceActions.lagda:271]:
```suggestion
governance action (\AgdaSymbol{:}\,\AgdaRecord{EnactState}) (see Figure~\ref{fig:enactment-types}).
```

File: src/Ledger/GovernanceActions.lagda [hunk 13]
```diff
@@ -181,22 +314,23 @@ instance
...
-  _⊢_⇀⦇_,ENACT⦈_ : EnactEnv → EnactState → GovAction → EnactState → Set
-\end{code}
-\begin{code}[hide]
-data _⊢_⇀⦇_,ENACT⦈_ where
-\end{code}
-\begin{code}
 
-  Enact-NoConf    : ⟦ gid ⟧ᵉ ⊢ s ⇀⦇ NoConfidence            ,ENACT⦈  record s { cc = nothing , gid }
-  Enact-NewComm   : ⟦ gid ⟧ᵉ ⊢ s ⇀⦇ NewCommittee new rem q  ,ENACT⦈ let
+The data type \verb!_⊢_⇀⦇_,ENACT⦈_! defines the transition relation for enacting a governance action.
```
Comment (WhatisRT) [inline src/Ledger/GovernanceActions.lagda:318]:
```suggestion
The relation \verb!_⊢_⇀⦇_,ENACT⦈_! is the transition relation for enacting a governance action.
```

**Review by WhatisRT — CHANGES_REQUESTED** [review]
There are also a lot of type annotations in the text that should be removed

File: src/Ledger/GovernanceActions.lagda [hunk 14]
```diff
@@ -1,5 +1,19 @@
...
+\item
+  a constitutional committee  (henceforth called \AgdaInductiveConstructor{CC})
+\item
+  a group of delegate representatives (henceforth called \AgdaInductiveConstructor{DReps})
+\item
+  the stake pool operators (henceforth called \AgdaInductiveConstructor{SPOs})
+\end{enumerate}
+Every governance action must be ratified by at least two of these three bodies using their on-chain \defn{votes}.
+The type of action and the state of the governance system determines which bodies must ratify it.
+Ratified actions are then \defn{enacted} on-chain, following a set of well-defined rules.
```
[the remark this answers is not in the document]
Comment (williamdemeo) [reply src/Ledger/GovernanceActions.lagda:16]: Again, this is probably quoted directly from the CIP... anyway, I'll make the suggested change. 👍🏼

File: src/Ledger/GovernanceActions.lagda [hunk 15]
```diff
@@ -103,7 +162,21 @@ module _ (pp : PParams) (ccThreshold' : Maybe ℚ) where
   threshold (ChangePParams x _)  = λ { CC → ccThreshold ; DRep → P5 x ; SPO → 0ℚ }
   threshold (TreasuryWdrl _)     = λ { CC → ccThreshold ; DRep → P6   ; SPO → 0ℚ }
   threshold Info                 = λ { CC → 2ℚ          ; DRep → 2ℚ   ; SPO → 2ℚ }
+\end{code}
+A governance action is an on-chain event that is triggered by a transaction, and there is a deadline after which
+the governance action cannot be enacted.
+\begin{itemize}
+\item
+  \defn{Ratification}. An action is said to be \defn{ratified} when it gathers enough votes in its favor
+  (according to the rules described in Section~\ref{sec:ratification}).
```
[the remark this answers is not in the document]
Comment (williamdemeo) [reply src/Ledger/GovernanceActions.lagda:171]: Correct, but it will exist once the following PR (#159) gets merged.

File: src/Ledger/GovernanceActions.lagda [hunk 16]
```diff
@@ -113,6 +186,23 @@ NeedsHash (ChangePParams _ _)  = GovActionID
...
+
+\subsection{Voting and ratification}
+\label{sec:voting-and-ratification}
+The data type \AgdaDatatype{Vote} represents the different voting options: \AgdaInductiveConstructor{yes}, \AgdaInductiveConstructor{no}, or \AgdaInductiveConstructor{abstain}.
+Each \AgdaField{vote} is recorded in a \AgdaRecord{GovVote} record along with the following data:
+\FieldTyped{gid}{GovActionID}, \FieldTyped{role}{GovRole}, \FieldTyped{credential}{Credential}, \FieldTyped{anchor}{Maybe~Anchor}.
+
+A \defn{governance action proposal} is recorded in a \AgdaRecord{GovProposal} record which includes fields for
+a return address (\FieldTyped{returnAddr}{RwdAddr}), the proposed governance action (\FieldTyped{action}{GovAction}),
+a hash of the previous governance action (\FieldTyped{prevAction}{NeedsHash action}), and an anchor. (See Figure~\ref{defs:governance-votes}.)
```
[the remark this answers is not in the document]
Comment (williamdemeo) [reply src/Ledger/GovernanceActions.lagda]: Although I prefer to write these "see Figure..." hints as separate sentences, it seems you prefer to make them part of the sentence that precedes them, so, for consistency, I'll change all occurrences to use the style you suggest, okay?

File: src/Ledger/GovernanceActions.lagda (hunk 9 above)
Comment (williamdemeo) [reply src/Ledger/GovernanceActions.lagda:116]: Ummm... not sure why we would do that, but okay.

**Comment by williamdemeo** [comment]
I think it looks pretty good now.  There's one unresolved reference which will be resolved as soon as we merge my next PR ( #159 ).

File: src/Ledger/GovernanceActions.lagda (hunk 16 above)
[the remark this answers is not in the document]
Comment (WhatisRT) [reply src/Ledger/GovernanceActions.lagda]: I don't necessarily mind having separate sentences for references (I usually prefer something integrated, but either way is fine), but then they shouldn't have parentheses. If you want parentheses, the main sentence should take care of punctuation. See for example https://www.utsa.edu/twc/documents/Parentheses.pdf.

**Comment by williamdemeo** [comment]
On Fri, Aug 4, 2023 at 3:43 AM Andre Knispel ***@***.***>
wrote:

> I don't necessarily mind having separate sentences for references (I
> usually prefer something integrated, but either way is fine), but then they
> shouldn't have parentheses. If you want parentheses, the main sentence
> should take care of punctuation. See for example
> https://www.utsa.edu/twc/documents/Parentheses.pdf.
>
From the document you linked to, "It’s also possible to have a
parenthetical that is its own sentence, not encased within a larger
sentence. In this case, it is punctuated like a full and independent
sentence."

**Comment by williamdemeo** [comment]
I don't think there are any outstanding change requests, but please correct me if I'm wrong.  Otherwise, LGTM. 👍🏼

File: src/Ledger/GovernanceActions.lagda [hunk 17]
```diff
@@ -61,7 +70,48 @@ data GovAction : Set where
...
+ >{\raggedright\arraybackslash}p{(\columnwidth - 2\tabcolsep) * \real{0.75}}@{}}
+\textbf{Action}  & \textbf{Description}\\
+\hline
+\endhead
+\AgdaInductiveConstructor{NoConfidence}            & a motion to create a \defn{state of no-confidence} in the current constitutional committee \\[10pt]
+\AgdaInductiveConstructor{NewCommittee}            & changes to the members of the constitutional committee and/or to its signature threshold and/or term limits \\[10pt]
+\AgdaInductiveConstructor{NewConstitution}         & a modification to the off-chain Constitution, recorded as an on-chain hash of the text document \\[10pt]
+\AgdaInductiveConstructor{TriggerHF}\footnotemark  & triggers a non-backwards compatible upgrade of the network; requires a prior software upgrade  \\[10pt]
+\AgdaInductiveConstructor{ChangePParams}           & a change to \textit{one or more} updatable protocol parameters, excluding changes to major protocol versions (``hard forks'')\\[10pt]
+\AgdaInductiveConstructor{TreasuryWdrl}            & movements from the treasury, sub-categorized into small, medium or large withdrawals (based on the amount of Lovelace to be withdrawn)\\[10pt]
```
Comment (WhatisRT) [inline src/Ledger/GovernanceActions.lagda:104]:
```suggestion
\AgdaInductiveConstructor{TreasuryWdrl}            & movements from the treasury \\[10pt]
```
The CIP is outdated here (and I have an open PR to fix it)

File: src/Ledger/GovernanceActions.lagda [hunk 18]
```diff
@@ -103,7 +162,21 @@ module _ (pp : PParams) (ccThreshold' : Maybe ℚ) where
...
+  \defn{Ratification}. An action is said to be \defn{ratified} when it gathers enough votes in its favor
+  (according to the rules described in Section~\ref{sec:ratification}).
+\item
+  \defn{Expiration}. An action that doesn't collect sufficient \AgdaField{yes} votes before its deadline is said to have \defn{expired}.
+\item
+  \defn{Enactment}. An action that has been ratified is said to be \defn{enacted} once it has been activated on the network.
+\end{itemize}
+See Section~\ref{sec:ratification} for more on the ratification process.
 
+\begin{code}
```
[the remark this answers is not in the document]
Comment (WhatisRT) [reply src/Ledger/GovernanceActions.lagda:164]: The `=` signs still aren't aligned (in the PDF)

File: src/Ledger/GovernanceActions.lagda [hunk 19]
```diff
@@ -131,23 +213,65 @@ record GovProposal : Set where
         prevAction  : NeedsHash action
         anchor      : Anchor
 \end{code}
-\begin{code}[hide]
+} %% end small
+\caption{Governance action proposals and votes}
+\label{defs:governance-votes}
+\end{figure*}
+The data type \AgdaDatatype{Vote} represents the different voting options: `yes', `no', or `abstain'.
```
Comment (WhatisRT) [inline src/Ledger/GovernanceActions.lagda:220]: These are constructors too.

**Review by WhatisRT — APPROVED** [review]
I found a few more small things, otherwise LGTM. Please fix those few open things before you merge.

**Comment by williamdemeo** [comment]
Oops, I merged this without fixing those small things.  I will make those changes part of the 145-Ratification (#159) PR.  (I've been spending too much time with git merging and rebasing today, so I don't want to do a proper reset in order to make these small changes part of #158 ...sorry.)

File: src/Ledger/GovernanceActions.lagda [hunk 20]
```diff
@@ -113,6 +171,30 @@ NeedsHash (ChangePParams _ _)  = GovActionID
...
+\end{figure*}
+Figure~\ref{fig:needshash-and-hashprotected-types} defines types that are used in
+ratification (for verifyPrev) where we check that the stored hash matches the one
+attached to the action we want to ratify.
+\begin{itemize}
+\item
+  \defn{Ratification}. An action is said to be \defn{ratified} when it gathers enough votes in its favor
+  (according to the rules described in Section~\ref{sec:ratification}).
+\item
+  \defn{Expiration}. An action that doesn't collect sufficient `yes' votes before its deadline is said to have \defn{expired}.
```
[the remark this answers is not in the document]
Comment (williamdemeo) [reply src/Ledger/GovernanceActions.lagda:189]: Okay, I'll change it back.  I had changed all yes, no and abstains from the Agda colored font to just quoted words, since I got the impression from your earlier comments that we want to make the docs more prosey and less codey... anyway, I'll change them all (or at least some of them) back.

**Comment by williamdemeo** [comment]
I've changed `yes`, `no`, and `abstain` from quoted plain-text words to AgdaInductiveConstructors in the GovernanceActions.lagda file.  If you'd like me to do the same in the Ratify.lagda file, please point that out when reviewing PR #159, okay?  Thanks!
